src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1293,8 +1293,9 @@ public Service getService(String type, String algorithm) {
         Service s = serviceMap.get(key);
         if (s == null) {
             s = legacyMap.get(key);
-            if (s != null && !s.isValid()) {
+            if (s != null && (!s.isValid() || !RestrictedSecurity.isServiceAllowed(s))) {
                 legacyMap.remove(key, s);
+                s = null;
```

Comment:
A comment would be helpful here:
```java
                // don't return invalid or disallowed legacy services
                s = null;
```

Reply:
Comment added.

In OpenJDK17 and earlier version, there is a method “ensureLegacyParsed()” to only add the valid service into legacyMap. And then go through the legacyMap to find the matching service, if no such service exists, then return null. But in the OpenJDK Next code, this “ensureLegacyParsed()” method is removed, instead, it changed from adding service to removing service. But after removing the invalid or not allowed service, it did not set “s = null”. So, this invalid or not allowed service “s” still be return. Added the “s = null” to avoid returning an invalid or disallowed service.